### PR TITLE
Fix read/write/is_allowed not called for dynamic attribute in async mode server

### PR DESCRIPTION
**Pull request by stanislaw55**
Hi,
this is PR with fix for #173.

Now it is a draft with (failing) tests only

### CODE REVIEW

**Comment by stanislaw55** [comment]
@ajoubertza, I have a problem with `Attr` class. It needs to get in its constructor one of tango types (`DevFloat`, `DevBoolean`) but types that I get from `typed_values` pytest fixture, are pure Python ones and tests are failing because of that

**Comment by stanislaw55** [comment]
Another thing, I wanted to test is_allowed method but I'm not sure how to do it. MY first idea was to use `state` fixture and bind output of is_allowed to state of the device but I'm not sure if that's good idea

**Comment by ajoubertza** [comment]
Thanks for starting on this!

> @ajoubertza, I have a problem with `Attr` class. It needs to get in its constructor one of tango types (`DevFloat`, `DevBoolean`) but types that I get from `typed_values` pytest fixture, are pure Python ones and tests are failing because of that

I see there are a few conversion method in `tango.utils`, although the one I'd prefer to use `tango.utils.get_tango_type()` fails.  So as a workaround, there is another implementation in the device_proxy module (and same again in pipe module) that can be used in the meantime.  Looks like we need to rationalise all these conversion methods!  

```python
In [1]: import tango, tango.test_utils                                                                                                                                

In [2]: [tango.device_proxy.__get_tango_type(t) for t in tango.test_utils.TYPED_VALUES]                                                                               
Out[2]: 
[tango._tango.CmdArgType.DevLong64,
 tango._tango.CmdArgType.DevDouble,
 tango._tango.CmdArgType.DevString,
 tango._tango.CmdArgType.DevBoolean,
 tango._tango.CmdArgType.DevVarLong64Array,
 tango._tango.CmdArgType.DevVarDoubleArray,
 tango._tango.CmdArgType.DevVarStringArray,
 tango._tango.CmdArgType.DevVarBooleanArray]
```

**Comment by ajoubertza** [comment]
> Another thing, I wanted to test is_allowed method but I'm not sure how to do it. MY first idea was to use `state` fixture and bind output of is_allowed to state of the device but I'm not sure if that's good idea

Maybe just set some variables if you hit the is_allowed method, and the actual attribute read/write methods.  Then assert on those?  Could try a class variable like [here](https://github.com/tango-controls/pytango/blob/c4df00fb5b77708174a3dc4286570a86656e4522/tests/test_server.py#L329).

**Comment by stanislaw55** [comment]
Hi @ajoubertza 
Thanks for help! Now I hope tests are good enough to catch proper bug

File: tests/test_server.py [hunk 1]
```diff
@@ -545,3 +545,41 @@ def cmd(self):
...
+            attr.set_value(self.attr_value)
+
+        def write_attr(self, attr):
+            self.attr_value = attr.get_write_value()
+
+        def is_attr_allowed(self, req_type):
+            return self._is_testattr_allowed
+
+    with DeviceTestContext(TestDevice) as proxy:
+        proxy._is_testattr_allowed = True
```
Comment (ajoubertza) [inline tests/test_server.py]: The `proxy` object is an instance of `DeviceProxy` connected to the device being tested.  The `DeviceTestContext` doesn't give us direct access to the Tango device it runs.  This means that line 577 isn't modifying `TestDevice._is_testattr_allowed` (or the instance variable for the device under test), but rather adding a new attribute to the `proxy` object. 

Add a static command or read/write attribute to `TestDevice` and use that to toggle the internal variable.

File: tests/test_server.py [hunk 2]
```diff
@@ -545,3 +545,41 @@ def cmd(self):
...
+        for value in values:
+            proxy.TestAttr = value
+            assert_close(proxy.attr, expected(value))
```
Comment (ajoubertza) [inline tests/test_server.py]: `proxy.attr` needs to be `proxy.TestAttr`.

File: tests/test_server.py [hunk 3]
```diff
@@ -545,3 +545,41 @@ def cmd(self):
...
+
+        proxy._is_testattr_allowed = False
+        for value in values:
+            with pytest.raises(DevFailed) as record:
```
Comment (ajoubertza) [inline tests/test_server.py]: If `record` isn't going to be used, it should be removed.

File: tests/test_server.py [hunk 4]
```diff
@@ -545,3 +545,41 @@ def cmd(self):
...
+                proxy.TestAttr = value
```
Comment (ajoubertza) [inline tests/test_server.py:589]: We should check that the "is_allowed" method also prevents reading:
```python
for value in values:
    ...
    with pytest.raises(DevFailed):
        proxy.TestAttr
```

**Review by ajoubertza — COMMENTED** [review]
That's looking better.  I've added some more suggestions.

**Comment by ajoubertza** [comment]
Thanks for the updates @stanislaw55 - its looking better.  I've resolved the comments you have addressed.

**Pull request closed without merging**